crates/turbo-tasks/src/task/concrete_task_input.rs

```diff
@@ -19,24 +19,49 @@ use crate::{
     registry, turbo_tasks, CellId, RawVc, TaskId, TraitType, ValueTypeId,
 };
 
+/// A reference to a piece of data with optional type information
 #[derive(Clone)]
-pub struct SharedReference(pub Option<ValueTypeId>, pub Arc<dyn Any + Send + Sync>);
+pub struct SharedReference<T: TypeData>(pub T, pub Arc<dyn Any + Send + Sync>);
```

Comment:
Instead of a `SharedReference<T>` I think it would be clearer if it would be `TypedSharedReference` and `SharedReference`.

`CellContent` would store a (untyped) `SharedReference`, since that's what cells store. We don't need an generic argument here. There is only one variant of content stored in cells.

`try_read_task_cell` need to be changed to not return `CellContent` but `Option<TypedSharedReference>` instead. It will add the type info.